### ISSUE TITLE
Extra empty line at the end of a block because that empty line is a comment

### ISSUE THREAD

**Issue by zimmski** [mgechev/revive#670]
(i am currently in the process of adding "revive" to our linter stack and i am going through the false-positives for us)

**Describe the bug**

```go
{
   some code

   // comment
}
```

leads to "extra empty line at the end of a block" even though it shouldn't.

**Expected behavior**

Such comments are meant for exactly being at that location. Marking the comment as a probem, is a false-positive for our source code.

**Comment by zimmski** [comment]
Same is true for 

```go
{
// comment
}
```

**Comment by git-hulk** [comment]
@chavacava I would like to handle this issue.

**Comment by git-hulk** [comment]
> Same is true for
> 
> ```go
> {
> // comment
> }
> ```

It's true since revive will check nothing when the block list is empty, so any empty line in the empty block will be ok(I think it can also regard as a bug).

**Comment by git-hulk** [comment]
For this case:
```
{
   some code

   // comment
}
```

The owner(key in `w.cmap`) of the comment line is `FuncDef` instead of the last node, so is between comment check is false. The workaround can be that records the current block's FuncDef and lookup the comment group with `FuncDef` node as well. How do you think about this solution?

Workaround from the user side can be:

```
{
   some code
   // comment
}```
then the owner of the comment line is last node instead of FuncDef in Go ast.

**Comment by chavacava** [comment]
Fixing the bug might require to stop using the `cmap` as it is currently used.
The way comments are attached to the AST nodes does not help us. 

We could just extract from `cmap` the list of positions (line, col) of all comments in the whole source file. Then, using that list, for each statement block we can check if there is an empty line (i.e. without an statement or a comment) just after/before the start/end of the open/close curly brace.

For example:
```go
{
   // this is an example
  println("Hello")

   // comment
}
```

From the above code we can get the list of positions of comments: `[(2,3),(5,3)]`
Then we check the block as follows:
1. Check the beginning of the block: Do the line just after the start of the block (line 2) is the position of the first statement OR the  position of a comment? Yes (line 2 is in the list of comments positions) then OK.
2. Check the end of the block: Do the line just before the end of the block (line 5) is the position of the last statement of the block OR the  position of a comment? Yes (line 5 is in the list of comments positions) then OK..

**Comment by chavacava** [comment]
@git-hulk please take a look into [this commit](https://github.com/mgechev/revive/commit/bf23317d8f65a77d2f4d72d4dfbbdd6782a242b5) as a guide on how the bug can be fixed

**Comment by git-hulk** [comment]
Cool, thanks for @chavacava guide.
